src/metabase/api/cloud_migration.clj

```diff
@@ -0,0 +1,245 @@
+                                                 #_resolved-query :default]
+  [_query-type model _parsed-args resolved-query]
+  (when (and (read-only-mode)
+             (not (read-only-mode-exceptions model))
+             (not *ignore-read-only-mode*))
+    (throw (ex-info (tru "Metabase is in read-only-mode mode!")
+                    {:status-code 403})))
+  resolved-query)
+
+
+;; Helpers
+
+(def ^:private terminal-states
+  #{:done :error :cancelled})
+
+(defn- cluster?
+  []
+  (>= (t2/count 'QRTZ_SCHEDULER_STATE) 2))
+
+(defn- progress-file-input-stream
+  "File input stream that calls on-percent-progress with current read progress as int from 0 to 100.
+  Does not call on-percent-progress for the same value twice. "
+  [^File file on-percent-progress]
+  (let [input-stream             (io/input-stream file)
+        length                   (.length file)
+        *bytes                   (atom 0)
+        on-percent-progress-memo (memoize on-percent-progress) ;; memoize so we don't repeat calls
+        add-bytes                #(on-percent-progress-memo (int (* 100 (/ (swap! *bytes + %) length))))
+        f                        (fn [ret]
+                                   (cond
+                                     ;; -1 is end of stream
+                                     (= -1 ret)  (on-percent-progress-memo 100)
+                                     (char? ret) (add-bytes 1)
+                                     (int? ret)  (add-bytes ret))
+                                   ret)]
+    (proxy [InputStream] []
+      (read
+       ([]
+        (f (.read input-stream)))
+        ([^bytes b]
+         (f (.read input-stream b)))
+        ([^bytes b off len]
+         (f (.read input-stream b off len))))
+      (close [] (.close input-stream)))))
+
+(defn- set-progress
+  "Attempt to set id to state and progress.
+  Throws if the migration has already reached a terminal state (e.g. cancelled).
+  This is the main cluster coordination mechanism for migrations, since any instance
+  can cancel the migration, not just the one that initiated it."
+  [id state progress]
+  (when (= 0 (t2/update! CloudMigration :id id :state [:not-in terminal-states]
+                         {:state state :progress progress}))
+    (throw (ex-info "Cannot update migration in terminal state" {:terminal true}))))
+
+(defn- migrate!
+  "Migrate this instance to Metabase Cloud.
+  Will exit early if migration has been cancelled in any cluster instance."
+  [{:keys [id upload_url]} & {:keys [retry?]}]
+  ;; dump-to-h2 starts behaving oddly if you try to dump repeatly to the same file
+  ;; in the same process.
+  (let [dump-file (io/file (str "cloud_migration_dump_" (random-uuid) ".mv.db"))
+        ;; Note: this will still set progress up to twice for each percent
+        ;; because e.g. 70 and 71 upload percent both map to 50+35 migration percent.
+        on-upload-progress #(set-progress id :upload (int (+ 50 (* 50 (/ % 100)))))]
```

Comment:
why we need to complecate this percentage tho? I'd prefer a simple `#(set-progress id :upload %)`